genesyscloud/util_files.go

```diff
@@ -2,119 +2,113 @@ package genesyscloud
-	var reader io.Reader
-	var file *os.File
-
-	_, err := os.Stat(path)
-	if err != nil {
-		_, err = url.ParseRequestURI(path)
-		if err == nil {
-			resp, err := http.Get(path)
-			if err != nil {
-				return nil, nil, err
-			}
-			if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
-				return nil, nil, fmt.Errorf("HTTP Error downloading file: %v", resp.StatusCode)
-			}
-			reader = resp.Body
-		} else {
-			return nil, nil, fmt.Errorf("Invalid file path or URL: %v", path)
-		}
-	} else {
-		file, err = os.Open(path)
-		if err != nil {
-			return nil, nil, err
-		}
-		reader = file
-	}
-
-	return reader, file, nil
+type S3Uploader struct {
+	reader        io.Reader
+	substitutions map[string]interface{}
+	headers       map[string]string
+	presignedUrl  string
+	client        http.Client
 }
 
-// Hash file content, used in stateFunc for "filepath" type attributes
-func hashFileContent(path string) string {
-	reader, file, err := downloadOrOpenFile(path)
-	if err != nil {
-		return err.Error()
-	}
-	if file != nil {
-		defer file.Close()
+func NewS3Uploader(reader io.Reader, substitutions map[string]interface{}, headers map[string]string, presignedUrl string) *S3Uploader {
+	c := &http.Client{}
+	s3Uploader := &S3Uploader{
+		reader:        reader,
+		substitutions: substitutions,
+		headers:       headers,
+		presignedUrl:  presignedUrl,
+		client:        *c,
 	}
 
-	hash := sha256.New()
-	if file == nil {
-		if _, err := io.Copy(hash, reader); err != nil {
-			return err.Error()
-		}
-	} else {
-		if _, err := io.Copy(hash, file); err != nil {
-			return err.Error()
-		}
-	}
-
-	return hex.EncodeToString(hash.Sum(nil))
+	log.Printf("%#v\n", s3Uploader)
+	return s3Uploader
 }
 
-// Read and upload input file path to S3 pre-signed URL
-func prepareAndUploadFile(filename string, substitutions map[string]interface{}, headers map[string]string, presignedUrl string) ([]byte, error) {
-	bodyBuf := &bytes.Buffer{}
-
-	reader, file, err := downloadOrOpenFile(filename)
-	if err != nil {
-		return nil, err
-	}
-	if file != nil {
-		defer file.Close()
-	}
-
-	_, err = io.Copy(bodyBuf, reader)
-	if err != nil {
-		return nil, fmt.Errorf("Failed to copy file content to the handler. Error: %s ", err)
-	}
-
+func (s *S3Uploader) substituteValues(bodyBuf *bytes.Buffer) {
 	// Attribute specific to the flows resource
-	if len(substitutions) > 0 {
+	if len(s.substitutions) > 0 {
 		fileContents := bodyBuf.String()
-		for k, v := range substitutions {
+		for k, v := range s.substitutions {
 			fileContents = strings.Replace(fileContents, fmt.Sprintf("{{%s}}", k), v.(string), -1)
 		}
 
 		bodyBuf.Reset()
 		bodyBuf.WriteString(fileContents)
 	}
+}
 
-	req, _ := http.NewRequest("PUT", presignedUrl, bodyBuf)
-	for key, value := range headers {
-		req.Header.Set(key, value)
+func (s *S3Uploader) Upload() ([]byte, error) {
+	bodyBuf := &bytes.Buffer{}
+
+	_, err := io.Copy(bodyBuf, s.reader)
```

Comment:
Notice the reader.  Originally we were passing in the filename and then looking up the file inside of the method.  Using a reader allows us to stub out the test case without having to have a file written out.